torch/nn/modules/module.py

```diff
@@ -189,17 +189,53 @@ def add_module(self, name, module):
             raise KeyError("module name can't be empty string \"\"")
         self._modules[name] = module
 
-    def _apply(self, fn):
+    def _apply(self, fn, change_params_inplace_cpu_cuda):
         for module in self.children():
-            module._apply(fn)
+            module._apply(fn, change_params_inplace_cpu_cuda)
+
+        def compute_should_move_tensor(tensor, tensor_applied):
+            # If the new tensor is still on the same device, we don't move
+            # the existing tensor (and we in-place update the existing tensor
+            # instead).
+            if tensor.device != tensor_applied.device:
+                # If the new tensor is on a different device, then we take
+                # `change_params_inplace_cpu_cuda` into account only if we are
+                # moving the model between CPU and CUDA. Otherwise, we always
+                # move the existing tensor.
+                if (tensor.is_cuda and tensor_applied.device == torch.device('cpu')) or
+                   (tensor.device == torch.device('cpu') and tensor_applied.is_cuda):
+                    return not change_params_inplace_cpu_cuda
+                else:
+                    return True
+            else:
+                return False
 
-        for param in self._parameters.values():
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                param_applied = fn(param.data)
+                if compute_should_move_tensor(param, param_applied):
+                    # Tensors stored in modules are graph leaves, and we don't want to
+                    # create copy nodes, so we have to use `with torch.no_grad():`
+                    with torch.no_grad():
+                        self.register_parameter(key, Parameter(param_applied, param.requires_grad))
+                        # Bump up the version counter of the original parameter, to invalidate
+                        # any previous references of it in the autograd graph.
+                        param._bump_version()
```

Comment:
Note to self: don't manually bump the version counter, because we should let the `fn()` handle version counter bumping. Also add a test to make sure an in-place `fn()` actually bumps the version of the new parameter, and an out-of-place `fn()` doesn't bump the version counter.